crates/http/src/spin.rs

```diff
@@ -1,231 +1,241 @@
-        };
-
-        Ok(())
-    }
-
-    fn params(uri: &Uri) -> Result<Vec<(String, String)>> {
-        match uri.query() {
-            Some(q) => Ok(url::form_urlencoded::parse(q.as_bytes())
-                .into_owned()
-                .collect::<Vec<_>>()),
-            None => Ok(vec![]),
-        }
-    }
-}
-
-fn contextualise_err(e: anyhow::Error) -> anyhow::Error {
-    if e.to_string()
-        .contains("failed to find function export `canonical_abi_free`")
-    {
-        e.context(
-            "component is not compatible with Spin executor - should this use the Wagi executor?",
-        )
-    } else {
-        e
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    #[test]
-    fn test_spin_header_keys() {
-        assert_eq!(
-            SpinHttpExecutor::prepare_header_key("SPIN_FULL_URL"),
-            "spin-full-url".to_string()
-        );
-        assert_eq!(
-            SpinHttpExecutor::prepare_header_key("SPIN_PATH_INFO"),
-            "spin-path-info".to_string()
-        );
-        assert_eq!(
-            SpinHttpExecutor::prepare_header_key("SPIN_RAW_COMPONENT_ROUTE"),
-            "spin-raw-component-route".to_string()
-        );
-    }
-}
+use crate::{
+    spin_http::{Method, SpinHttp},
+    ExecutionContext, HttpExecutor, RuntimeContext,
+};
+use anyhow::Result;
+use async_trait::async_trait;
+use http::Uri;
+use hyper::{Body, Request, Response};
+use spin_engine::io::{ModuleIoRedirects, ModuleIoRedirectsTypes};
+use std::{net::SocketAddr, str, str::FromStr};
+use tokio::task::spawn_blocking;
+use tracing::log;
+use wasmtime::{Instance, Store};
+
+#[derive(Clone)]
+pub struct SpinHttpExecutor;
+
+#[async_trait]
+impl HttpExecutor for SpinHttpExecutor {
+    async fn execute(
+        &self,
+        engine: &ExecutionContext,
+        component: &str,
+        base: &str,
+        raw_route: &str,
+        req: Request<Body>,
+        _client_addr: SocketAddr,
+        follow: bool,
+    ) -> Result<Response<Body>> {
+        log::trace!(
+            "Executing request using the Spin executor for component {}",
+            component
+        );
+
+        let mut mior = if let ModuleIoRedirectsTypes::FromFiles(clp) = engine.config.module_io_redirects.clone() {
+            ModuleIoRedirects::new_from_files(
+                clp.stdin_pipe.map(|inp| inp.0),
+                clp.stdout_pipe.map(|oup| oup.0),
+                clp.stderr_pipe.map(|erp| erp.0),
+            )
+        } else {
+            ModuleIoRedirects::new(follow)
+        };
+
+        let (pipes, read_handles) = (mior.pipes.take(), mior.read_handles.take());
+
+        let (store, instance) = engine.prepare_component(component, None, pipes, None, None)?;
+
+        let resp_result = Self::execute_impl(store, instance, base, raw_route, req)
+            .await
+            .map_err(contextualise_err);
+
+        let log_result =
+            engine.save_output_to_logs(read_handles.unwrap().read(), component, true, true);

```

Comment:
I'm using an `Option` there to be able to `take` that value from `mior`. That said, I'll use the `if let` approach then (:

Reply:
I guess I am not quite understanding why the `take` might return a `None`.  Is that possible?  If so, what does it represent?

Reply:
If I look at the `ModuleIoRedirects` constructors, both of them place a value in the `read_handles` field.  So the `take` can never return a None.  So why not make `read_handles` a plain value instead of an Option?  I feel like I'm missing something, sorry.

Reply:
Like if I make the ModuleIoRedirect fields plain values, and I change the executor line to:

```
let (pipes, read_handles) = (mior.pipes, mior.read_handles);
```

then it all compiles, at least, but maybe it's not representing the cases you need it to represent?

Reply:
You are right. I remembered that I added that `Option` to fix a shared reference problem I was running into, but that doesn't seem to be an issue anymore — I'll make them plain values!

Reply:
done ✅